### PR TITLE
Per project versioning

### PR DESCRIPTION
**Pull request by arkpar**

### CODE REVIEW

**Comment by arkpar** [comment]
https://github.com/ethereum/webthree-helpers/pull/18
https://github.com/ethereum/libweb3core/pull/6
https://github.com/ethereum/libethereum/pull/19
https://github.com/ethereum/mix/pull/75
https://github.com/ethereum/solidity/pull/79
https://github.com/ethereum/webthree/pull/41
https://github.com/ethereum/alethzero/pull/63
These PRs go together

File: cmake/EthDependencies.cmake [hunk 1]
```diff
@@ -179,17 +67,26 @@ elseif (UNIX)
...
 
+include_directories(BEFORE "${PROJECT_BINARY_DIR}/include")
+
 function(eth_use TARGET REQUIRED)
+	if (NOT TARGET ${TARGET})
+		message(FATAL_ERROR "eth_use called for non existing target ${TARGET}")
+	endif()
+
+	if (TARGET ${PROJECT_NAME}_BuildInfo.h)
+		add_dependencies(${TARGET} ${PROJECT_NAME}_BuildInfo.h)
```
Comment (debris) [inline cmake/EthDependencies.cmake:87]: nice hack :+1:

File: cmake/EthDependencies.cmake [hunk 2]
```diff
@@ -198,7 +95,11 @@ function(eth_use TARGET REQUIRED)
 			list(GET MODULE_PARTS 1 MODULE_SUB)
 		endif()
 		# TODO: check if file exists if not, throws FATAL_ERROR with detailed description
-		include(Use${MODULE_MAIN})
-		eth_apply(${TARGET} ${REQUIRED} ${MODULE_SUB})
+		get_target_property(TARGET_APPLIED ${TARGET} TARGET_APPLIED_${MODULE_MAIN}_${MODULE_SUB})
```
Comment (debris) [inline cmake/EthDependencies.cmake:98]: :+1:

File: cmake/EthOptions.cmake [hunk 3]
```diff
@@ -34,6 +38,10 @@ macro(configure_project)
 		add_definitions(-DETH_VMTRACE)
 	endif ()
 
+	if (JSCONSOLE)
+		add_definitions(-DETH_JSCONSOLE)
```
Comment (debris) [inline cmake/EthOptions.cmake:42]: I'm not sure if build without jsconsole is still possible. I also removed `JSCONSOLE` variable from default values.

File: cmake/EthOptions.cmake [hunk 4]
```diff
@@ -45,31 +53,64 @@ macro(configure_project)
...
+if (SUPPORT_TOOLS)
 	message("-- TOOLS            Build tools                              ${TOOLS}")
+endif()
+if (SUPPORT_ETHASHCL)
 	message("-- ETHASHCL         Build OpenCL components                  ${ETHASHCL}")
+endif()
+if (SUPPORT_EVMJIT)
 	message("-- EVMJIT           Build LLVM-based JIT EVM                 ${EVMJIT}")
+endif()
+if (SUPPORT_SOLIDITY)
```
Comment (debris) [inline cmake/EthOptions.cmake]: where `SUPPORT_XXX` variables come from?

File: cmake/UseDev.cmake [hunk 5]
```diff
@@ -9,17 +9,11 @@ function(eth_apply TARGET REQUIRED SUBMODULE)
...
-
-	target_include_directories(${TARGET} BEFORE PUBLIC ${CMAKE_BUILD_DIR})
-	target_include_directories(${TARGET} BEFORE PUBLIC ${CMAKE_BINARY_DIR})
-	target_include_directories(${TARGET} BEFORE PUBLIC ${DEV_BUILD_DIR})
-	endif()
+	#if (${SUBMODULE} STREQUAL "buildinfo")
+	#target_include_directories(${TARGET} BEFORE PUBLIC ${CMAKE_BUILD_DIR})
+	#target_include_directories(${TARGET} BEFORE PUBLIC ${CMAKE_BINARY_DIR})
+	#target_include_directories(${TARGET} BEFORE PUBLIC ${DEV_BUILD_DIR})
+	#endif()
```
Comment (debris) [inline cmake/UseDev.cmake]: these lines can be removed

File: cmake/EthOptions.cmake [hunk 6]
```diff
@@ -17,6 +17,10 @@ macro(configure_project)
 	eth_default_option(EVMJIT OFF)
 	eth_default_option(SOLIDITY ON)
 
+	foreach(FEATURE ${ARGN})
+		set(SUPPORT_${FEATURE} TRUE)
```
Comment (debris) [inline cmake/EthOptions.cmake]: `FEATURE` always overwrites option passed to cmake by user.

Comment (debris) [inline cmake/EthOptions.cmake:21]: what is the purpose of feature? Is it only reducing number of logs?

File: cmake/UseWeb3.cmake [hunk 7]
```diff
@@ -19,7 +19,7 @@ function(eth_apply TARGET REQUIRED SUBMODULE)
 	endif()
 
 	if (${SUBMODULE} STREQUAL "web3jsonrpc")
-		eth_use(${TARGET} ${REQUIRED} Web3::webthree JsonRpc::Server)
+		eth_use(${TARGET} ${REQUIRED} Mhd Web3::webthree JsonRpc::Server)
```
Comment (debris) [inline cmake/UseWeb3.cmake:22]: Can `Mhd` be a part of `JsonRpc::Server` ?

File: cmake/UseJsonRpc.cmake [hunk 8]
```diff
@@ -1,8 +1,18 @@
...
 
 	if (${SUBMODULE} STREQUAL "Server")
+		get_property(DISPLAYED GLOBAL PROPERTY ETH_JSONRPCSTUB_DISPLAYED)
+		if (NOT DISPLAYED)
+			set_property(GLOBAL PROPERTY ETH_JSONRPCSTUB_DISPLAYED TRUE)
+			message(STATUS "jsonrpcstub location    : ${ETH_JSON_RPC_STUB}")
+		endif()
 		target_include_directories(${TARGET} SYSTEM PUBLIC ${JSON_RPC_CPP_INCLUDE_DIRS})
 		target_link_libraries(${TARGET} ${JSON_RPC_CPP_SERVER_LIBRARIES})
 		target_link_libraries(${TARGET} ${MHD_LIBRARIES})
```
Comment (debris) [inline cmake/UseJsonRpc.cmake]: It should be replaced with `UseMhd.cmake`

File: cmake/EthOptions.cmake (hunk 6 above)
Comment (arkpar) [inline cmake/EthOptions.cmake:21]: yep

File: cmake/UseWeb3.cmake (hunk 7 above)
Comment (arkpar) [inline cmake/UseWeb3.cmake:22]: It is also used in alethzero independently

File: cmake/EthOptions.cmake (hunk 3 above)
Comment (arkpar) [inline cmake/EthOptions.cmake:42]: looks like this is still required

**Pull request merged**